### PR TITLE
Use RAII to clear

### PR DESCRIPTION
**Pull request by xlqian**
When no solution is found or an exception is thrown, the structure in Algo should be cleared...

### CODE REVIEW

File: asgard/handler.cpp [hunk 1]
```diff
@@ -142,6 +142,25 @@ make_valhalla_locations_from_projected_locations(const std::vector<midgard::Poin
...
+    Finally(const Finally&) = delete;
+    Finally& operator=(const Finally&) = delete;
+    Finally& operator=(Finally&&) = delete;
+    ~Finally() {
+        t();
+    };
+};
+
+template<typename T>
+Finally<T> make_finally(T t) {
```
Comment (AzHicham) [inline asgard/handler.cpp:160]: Maybe use this code instead ?
template <typename T>
Finally<T> make_finally(T&& t)
{
    return { std::forward<T>(t) };
}

File: asgard/handler.cpp [hunk 2]
```diff
@@ -142,6 +142,25 @@ make_valhalla_locations_from_projected_locations(const std::vector<midgard::Poin
...
+struct Finally {
+    T t;
+    explicit Finally(T t) : t(t){};
+    Finally() = delete;
+    Finally(Finally&& f) = default;
+    Finally(const Finally&) = delete;
+    Finally& operator=(const Finally&) = delete;
+    Finally& operator=(Finally&&) = delete;
+    ~Finally() {
+        t();
```
Comment (AzHicham) [inline asgard/handler.cpp:155]: Maybe add try/catch if f can throw ?

File: asgard/handler.h [hunk 3]
```diff
@@ -48,6 +48,15 @@ struct Handler {
                                                       const valhalla::Location& destination,
                                                       const std::string& mode);
 
+    void clear() {
+        if (graph.OverCommitted()) { graph.Clear(); }
```
Comment (benoit-bst) [inline asgard/handler.h:52]: why you need _RAII_ to call clear functions ?

Comment (AzHicham) [reply asgard/handler.h:52]: To make sure we call clear function in all cases : 
- solution found
- no solution found
- Exception thrown

**Review by benoit-bst — APPROVED** [review]

**Pull request merged**